### PR TITLE
bugfix: #31268

### PR DESCRIPTION
**Pull request by shoryukenn**
issuse: https://github.com/flutter/flutter/issues/31268

Specify the link options in both modes explicitly, adapting different codes for different options.

### CODE REVIEW

**Comment by shoryukenn** [comment]
/subsystem:windows/conosle is not tied to debug/release.

About console:
1. The AllocConsole/AttachConsole allows the current application to have a new command line window.
2. You can redirect stdout to any console (like freopen) via AttachConsole(ATTACH_PARENT_PROCESS or pid).

Usually, we use the second, which means that example.vcxproj has /subsystem:windows by default. In the main function, we use AttachConsole(ATTACH_PARENT_PROCESS) to redirect stdout to the parent process (such as visual studio debug output or flutter run --release).

**Comment by shoryukenn** [comment]
I modified the logic, the code is from Chromium, the project property is changed to /subsystem:windows, and the console--force-console is used to control whether to create the console. At the same time, if the parent process has a console, it will be attached directly. I think this is more consistent, what do you think?

File: example/windows/flutter_embedder_example.cpp [hunk 1]
```diff
@@ -44,9 +55,88 @@ std::string GetExecutableDirectory() {
...
+}
+
 }  // namespace
 
-int main(int argc, char **argv) {
+
+int APIENTRY wWinMain(HINSTANCE instance, HINSTANCE prev, wchar_t * lpCmdLine, int) {
+
+  // 1. if parent or current process has console, attach it.
+  // 2. if --force-console exist && parent or current process have not conole,
```
[the remark this answers is not in the document]
Comment (shoryukenn) [reply example/windows/flutter_embedder_example.cpp]: ok, will done.  I think your opinion is correct. According to our discussion, we only need one or two lines of code.

**Comment by shoryukenn** [comment]
@stuartmorgan ,There is a scenario we have to confirm: 
What do we expect when Visual Studio debug example.exe?

File: example/windows/flutter_embedder_example.cpp [hunk 2]
```diff
@@ -46,7 +46,14 @@ std::string GetExecutableDirectory() {
 
 }  // namespace
 
-int main(int argc, char **argv) {
+int APIENTRY wWinMain(HINSTANCE instance, HINSTANCE prev, wchar_t *, int) {
+
+  if (AttachConsole(ATTACH_PARENT_PROCESS)) {
+    FILE *p_fCout = nullptr;
```
[the remark this answers is not in the document]
Comment (shoryukenn) [reply example/windows/flutter_embedder_example.cpp]: done

**Comment by shoryukenn** [comment]
> I tried this locally, and it completely broke `flutter run`. It's not attaching to the observatory, so things like `q` and `r` don't do anything. It's critical that the Flutter tooling continue to work.
> 
> > What do we expect when Visual Studio debug example.exe?
> 
> Can you elaborate on the question? I'm not sure what specific behavior you mean.

I think when I use WinDBG or VS(f5) to start FDE.exe, generating a Console will be more friendly to developers.  So do we need to add a logic ? such as:
If (::IsDebuggerPresent()) {
   AllocConsole();
}

**Comment by stuartmorgan** [comment]
> I think when I use WinDBG or VS(f5) to start FDE.exe, generating a Console will be more friendly to developers.

I see; I wasn't at my Windows machine, and had forgotten that VS doesn't itself act as a console for the running application (as compared to Xcode on macOS). I agree that if we can easily and reliably detect that case, showing a console would be helpful.

**Comment by shoryukenn** [comment]
> > I think when I use WinDBG or VS(f5) to start FDE.exe, generating a Console will be more friendly to developers.
> 
> I see; I wasn't at my Windows machine, and had forgotten that VS doesn't itself act as a console for the running application (as compared to Xcode on macOS). I agree that if we can easily and reliably detect that case, showing a console would be helpful.

DONE.

File: example/windows/flutter_embedder_example.cpp [hunk 3]
```diff
@@ -46,7 +46,16 @@ std::string GetExecutableDirectory() {
 
 }  // namespace
 
-int main(int argc, char **argv) {
+int APIENTRY wWinMain(HINSTANCE instance,
+                      HINSTANCE prev,
+                      wchar_t *command_line,
+                      int show_command) {
+  // attach to console of 'flutter run'
```
Comment (stuartmorgan) [inline example/windows/flutter_embedder_example.cpp]: Nit: s/attach/Attach/, and add missing period at the end. The style guide requires that comments be punctuated as sentences.

Comment (shoryukenn) [reply example/windows/flutter_embedder_example.cpp]: > I don't currently have access to my Windows machine. Did you confirm that this version works correctly with `flutter run`, attaching to the observatory and accepting/processing key input to trigger things like hot reload?

yes, I'm sure. It works fine in both scenarios (flutter run or VS debugger).
WXWorkCapture_15671489902245
WXWorkCapture_1567149044653

Comment (shoryukenn) [reply example/windows/flutter_embedder_example.cpp]: > Nit: s/attach/Attach/, and add missing period at the end. The style guide requires that comments be punctuated as sentences.

DONE.

**Pull request merged**